vllm/worker/model_input.py

```diff
@@ -0,0 +1,288 @@
+                    clone_kwargs[field.name] = val
+            clone_kwargs = cls._get_init_kwargs(**clone_kwargs)
+
+        kwargs = cls._get_init_kwargs(**kwargs)
+        return cls(**clone_kwargs, **kwargs)
+
+    def replace(self: T, **kwargs) -> T:
+        """
+        Replace current fields with fields in kwargs.
+        """
+        valid_kwargs = self.__class__._get_init_kwargs(**kwargs)
+        return dataclasses.replace(self, **valid_kwargs)
+
+    def as_broadcastable_tensor_dict(
+            self) -> Dict[str, Union[int, torch.Tensor]]:
+        """
+        Extract broadcastable fields. Override for fields that require some
+        custom deserialization.
+        """
+        tensor_dict: Dict[str, Union[int, torch.Tensor]] = {}
+        for field in self.broadcastable_fields:
+            val = getattr(self, field, None)
+            if val is not None:
+                tensor_dict[field] = val
+
+        return tensor_dict
+
+
+@dataclasses.dataclass(frozen=True)
+class CPUModelInput(ModelInput):
+    """
+    Used by the CPUModelRunner.
+    """
+    num_seq_groups: Optional[int] = None
+    blocks_to_copy: Optional[torch.Tensor] = None
+
+    input_tokens: Optional[torch.Tensor] = None
+    input_positions: Optional[torch.Tensor] = None
+    multi_modal_kwargs: Optional[Dict[str, torch.Tensor]] = None
+
+    attn_metadata: Optional["AttentionMetadata"] = None
+    sampling_metadata: Optional["SamplingMetadata"] = None
+
+    @property
+    def broadcastable_fields(self) -> Tuple[str, ...]:
+        return (
+            "num_seq_groups",
+            "blocks_to_copy",
+            "input_tokens",
+            "input_positions",
+            "multi_modal_kwargs",
+        )
+
+    @classmethod
+    def _get_init_kwargs(  # type: ignore
+            cls, **kwargs) -> Dict[str, Any]:
+        kwargs = _init_attn_metadata_from_kwargs(**kwargs)
+        kwargs = _init_sampling_metadata_from_kwargs(**kwargs)
+        return super()._get_init_kwargs(**kwargs)
+
+    def as_broadcastable_tensor_dict(
+            self) -> Dict[str, Union[int, torch.Tensor]]:
+        tensor_dict = super().as_broadcastable_tensor_dict()
+        _add_attn_metadata_broadcastable_dict(tensor_dict, self.attn_metadata)
+        _add_sampling_metadata_broadcastable_dict(tensor_dict,
+                                                  self.sampling_metadata)
+        return tensor_dict
+
+
+@dataclasses.dataclass(frozen=True)
+class GPUModelInput(ModelInput):
+    """
+    This base class contains metadata needed for the base model forward pass
+    but not metadata for possible additional steps, e.g., sampling. Model
+    runners that run additional steps should subclass this method to add
+    additional fields.
+    """
+    num_seq_groups: Optional[int] = None
+    blocks_to_swap_in: Optional[torch.Tensor] = None
+    blocks_to_swap_out: Optional[torch.Tensor] = None
+    blocks_to_copy: Optional[torch.Tensor] = None
+
```

Comment:
I feel this is abstraction leakage. Ideally we should have `WorkerInput` and `ModelInput`, and `Worker` should not peak into `ModelInput`.